Sources/Testing/Polling/Polling.swift

```diff
@@ -0,0 +1,293 @@
+        kind: .expectationFailed(expectation),
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    case .timedOutWithoutRunning:
+      expectation.isPassing = false
+      Issue(
+        kind: .expectationFailed(expectation),
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    case .finished:
+      return __checkValue(
+        await recorder.lastValue,
+        expression: expression,
+        comments: comments,
+        isRequired: isRequired,
+        sourceLocation: sourceLocation
+      )
+    case .failed:
+      return __checkValue(
+        nil,
+        expression: expression,
+        comments: comments,
+        isRequired: isRequired,
+        sourceLocation: sourceLocation
+      )
+    case .cancelled:
+      Issue(
+        kind: .system,
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    }
+
+    return .failure(ExpectationFailedError(expectation: expectation))
+  }
+
+  /// A type to record the last value returned by a closure returning an optional
+  /// This is only used in the `#require(until:)` macro returning an optional.
+  private actor Recorder<R: Sendable> {
+    var lastValue: R?
+
+    /// Record a new value to be returned
+    func record(value: R) {
+      self.lastValue = value
+    }
+  }
+
+  /// The result of polling expressions
+  private enum PollResult {
+    /// The polling timed out, and the expression had run at least once.
+    case timedOut
+    /// The polling timed out, but the expression had not finished running in
+    /// that time.
+    case timedOutWithoutRunning
+    /// The expression exited early, and we will report a success status.
+    case finished
+    /// The expression returned false under PollingBehavior.passesAlways
+    case failed
+    /// The polling was cancelled before polling could finish
+    case cancelled
+  }
+
+  /// The poll manager.
+  ///
+  /// This function contains the logic for continuously polling an expression,
+  /// as well as the logic for cancelling the polling once it times out.
+  ///
+  /// - Parameters:
+  ///   - expression: An expression to continuously evaluate
+  ///   - behavior: The polling behavior to use
+  ///   - timeout: How long to poll for unitl the timeout triggers.
+  /// - Returns: The result of this polling.
+  private static func poll(
+    expression: @escaping @Sendable () async -> Bool,
+    behavior: PollingBehavior,
+    timeout: Duration
+  ) async -> PollResult {
+    let pollingProcessor = PollingProcessor(behavior: behavior)
+    return await withTaskGroup { taskGroup in
+      taskGroup.addTask {
+        do {
+          try await Task.sleep(for: timeout)
+        } catch {}
+        // Task.sleep will only throw if it's cancelled, at which point this
+        // taskgroup has already returned and we don't care about the value
+        // returned here.
+        return await pollingProcessor.didTimeout()
+      }
```

Comment:
Per [discussion in forum](https://forums.swift.org/t/pitch-polling-expectations/79866/15), this is unsafe & needs to be rethought.